### ISSUE TITLE
v3 gradual slowdown

### ISSUE THREAD

**Issue by endlisnis** [MetPX/sarracenia#440]
I noticed a problem this morning with my v3 instance.  It was more than 20 minutes behind on the feed; and I could see from the logs that it would sometimes just sit there for 10 minutes at a time between downloads (even though I know there are many files per minute that it should be downloading).  I restarted the service and it quickly caught up; but the instance did not respond during the shutdown request:
```
$ ~/.local/bin/sr3 restart subscribe/rolf
restarting: sending SIGTERM .... ( 4 ) Done
Waiting 1 sec. to check if 4 processes stopped (try: 0)
Waiting 2 sec. to check if 4 processes stopped (try: 1)
Waiting 4 sec. to check if 4 processes stopped (try: 2)
Waiting 8 sec. to check if 4 processes stopped (try: 3)
Waiting 16 sec. to check if 4 processes stopped (try: 4)
doing SIGKILL this time
signal_pid( 2528642, SIGKILL )


.signal_pid( 2528644, SIGKILL )
.signal_pid( 2528645, SIGKILL )
.signal_pid( 2528643, SIGKILL )
.Done
Waiting again...
All stopped after KILL
.( 4 ) Done
```

I was running 4 instances at the time, and all 4 instances had similar patterns in their log files:
[subscribe_rolf_01.log.gz](https://github.com/MetPX/sarracenia/files/7835137/subscribe_rolf_01.log.gz)

**Comment by petersilva** [comment]
This is super interesting... what I see in the log, general trend:
* log starts is for today (2022-01-09)
* starts with 00h09 with each message nearly 1700 seconds, or 28 minutes of delay in download.
* lag keeps increasing adding another minute (29 total) before you restart @7h49.
* on restart, the lag starts dropping rapidly...  within 8 minutes... @7h57, it is caught up (0.5 seconds of lag.)
* then it wanders between <1 and 4 seconds... which looks normal.
* continues that way until end of log 08h33.  

Periodic "housekeeping intervals" are supposed to happen every 'housekeeping' interval, which is, by default, 5 minutes.
The actual intervals are all over the map... sample interval from beginning of the file: 09,23,37,50,03,07,21,34,48 ...  Now it only checks to run housekeeping between transfer batches... so it looks like individual batches are taking 20 minutes...

The processing of messages/files is done in batches, and you see a after_accept messages happen @ 0h09, then it is quiet for 15 minutes (until 0h23) it is downloading all those files.  It is not that it was sitting there doing nothing, it's that the success log is only happenning after the entire batch is complete.  

It is the transfers themselves from the datamart that are pathologically slow at those times... so the built-in transfers are slowing down... guesses:
* memory fragmentation?
* memory leak?

hmm... I'm going to look into some periodic plugins to report memory usage.  in the meantime,  you can restart every couple of days with a cron job, and that might prevent a recurrence, or if you are willing to tolerate a recurrence, I will provide a memory reporter, and We already have a bug outstanding where periodic processing is supposed to restart when memory leaks are detected ( #395 ) 

conclusion:

* You can program in a restart every few days to avoid this problem as a work-around.
* We will produce some memory management plugins and let you know when they are ready.
* fwiw, *batch 10* setting... smaller batches might help you see log messages closer to when the events happen.

**Comment by endlisnis** [comment]
[subscribe_rolf_01.log.2022-01-08.gz](https://github.com/MetPX/sarracenia/files/7835597/subscribe_rolf_01.log.2022-01-08.gz)

Here's yesterday's log file that shows the slowdown start in the evening.

**Comment by endlisnis** [comment]
It happened again: 
[subscribe_rolf_01.log.2022-01-15.gz](https://github.com/MetPX/sarracenia/files/7877722/subscribe_rolf_01.log.2022-01-15.gz)

You can ignore the 3am slowdown, that is caused by some local processes.  But at around 9:00am, it got very slow again and only sped up after I restarted it.

**Comment by petersilva** [comment]
looking at the log, exactly as you say... I'd like some memory information to be in the housekeeping reports... still waiting on #395

**Comment by endlisnis** [comment]
There is something else going on here too.  No matter how slow it gets, it never falls more than about 1770 seconds behind; and if I get an outage longer than ~1770 seconds, it resumes about 1770 seconds out (instead of the 4h that I have in my config).

It seems as if the queue is deleting messages after around 1770 seconds.

**Comment by petersilva** [comment]
bunch of different possibilities:

* might be a setting on our side. I think if a queue gets > 100,000 we start refusing to queue to it... to save our broker. 
  (@junhu3 what is the setting on dd.weather.gc.ca? ) 

how many updates do you get per hour approx?

**Comment by junhu3** [comment]
max-length:10000 for anonymous queues

**Comment by petersilva** [comment]
only ten thousand... yeah that makes more sense... it would do that...

**Comment by endlisnis** [comment]
I get ~20k items per hour, which makes the 1700 second limit make sense.

**Comment by petersilva** [comment]
fwiw @MagikEh is making great progress on #395 ... I think it will solve this.

**Comment by petersilva** [comment]
now that #395 is merged... @endlisnis you should update your configuration and stop/start. With the new code if the memory usage gets too large, it should restart on it´s own, so you don´t need to do anything.

**Comment by endlisnis** [comment]
I updated my clients.  Let's wait a week and see if I have this problem again (I was having it on average about once every 2 days).

**Comment by endlisnis** [comment]
I also (in the above log) get a bunch of messages like "download incomplete download only 9452 of expected 9455 bytes for 2022-02-23-1740-CVYD-AUTO-minute-swob.xml" when I restart the clients.

**Comment by petersilva** [comment]
There were lots of problems with internal systems yesterday.  The feeds to to DMZ systems were affected.
I don't think anything yesterday is representative of a problem with your feed.

**Comment by petersilva** [comment]
Is it back to normal today?

**Comment by endlisnis** [comment]
Yes, it seems fine today.

**Comment by endlisnis** [comment]
I've been suffering problems over the last few days that also cause my client to fall way behind.  I restarted them about 15 minutes ago and they quickly caught up.  Here is a log from instance 1:
[subscribe_rolf_01.log.gz](https://github.com/MetPX/sarracenia/files/8337477/subscribe_rolf_01.log.gz)

**Comment by petersilva** [comment]
great log... It isn't clear when you restarted... It doesn't look like it logged anything on shutdown or startup.
from your description and the log, I'm guessing 20:26?



summary of what is in the log
---------------------------------------
it started slowing here:

```
2022-03-23 12:57:56,825 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 3.00 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CVQZ/2022-03-23-1657-CVQZ-AUTO-minute-swob.xml
2022-03-23 12:57:56,825 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 3.00 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWAQ/2022-03-23-1657-CWAQ-AUTO-minute-swob.xml
2022-03-23 12:57:56,825 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.96 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWZE/2022-03-23-1657-CWZE-AUTO-minute-swob.xml
2022-03-23 12:57:56,825 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.96 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWKH/2022-03-23-1657-CWKH-AUTO-minute-swob.xml
2022-03-23 12:57:56,825 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.45 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWBM/2022-03-23-1657-CWBM-AUTO-minute-swob.xml
2022-03-23 12:59:17,937 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CXKI/2022-03-23-1657-CXKI-AUTO-minute-swob.xml
2022-03-23 12:59:17,937 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CXOX/2022-03-23-1657-CXOX-AUTO-minute-swob.xml
2022-03-23 12:59:17,937 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CXPT/2022-03-23-1657-CXPT-AUTO-minute-swob.xml
2022-03-23 12:59:17,937 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CXKE/2022-03-23-1656-CXKE-AUTO-minute-swob.xml
2022-03-23 12:59:17,937 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CWEW/2022-03-23-1657-CWEW-AUTO-minute-swob.xml
2022-03-23 12:59:17,937 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CVQZ/2022-03-23-1657-CVQZ-AUTO-minute-swob.xml
2022-03-23 12:59:17,938 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CWAQ/2022-03-23-1657-CWAQ-AUTO-minute-swob.xml
2022-03-23 12:59:17,938 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CWZE/2022-03-23-1657-CWZE-AUTO-minute-swob.xml
2022-03-23 12:59:17,938 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CWKH/2022-03-23-1657-CWKH-AUTO-minute-swob.xml
2022-03-23 12:59:17,938 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CWBM/2022-03-23-1657-CWBM-AUTO-minute-swob.xml
2022-03-23 12:59:20,440 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 61.79 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWDV/2022-03-23-1657-CWDV-AUTO-minute-swob.xml
2022-03-23 12:59:20,440 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 61.79 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CAMS/2022-03-23-1657-CAMS-AUTO-minute-swob.xml
2022-03-23 12:59:20,440 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 61.60 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CVXC/2022-03-23-1657-CVXC-AUTO-minute-swob.xml
2022-03-23 12:59:20,440 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 62.87 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWPN/2022-03-23-1657-CWPN-AUTO-minute-swob.xml
```

then was stable for a little while until

```

022-03-23 12:59:20,452 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 61.10 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CVOQ/2022-03-23-1657-CVOQ-AUTO-minute-swob.xml
2022-03-23 12:59:20,452 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 61.10 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWKX/2022-03-23-1657-CWKX-AUTO-minute-swob.xml
2022-03-23 12:59:20,452 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 61.10 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWOK/2022-03-23-1657-CWOK-AUTO-minute-swob.xml
2022-03-23 12:59:20,452 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 701.81 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CPUN/2022-03-23-1646-CPUN-AUTO-minute-swob.xml
2022-03-23 12:59:20,452 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 701.81 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWDJ/2022-03-23-1646-CWDJ-AUTO-minute-swob.xml
2022-03-23 12:59:20,452 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 701.81 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CVSJ/2022-03-23-1646-CVSJ-AUTO-minute-swob.xml

```

then it got gradually worse, until the worst it got was here:

```

2022-03-23 15:50:06,217 [INFO] sarracenia.flowcb.log housekeeping_stats lag: average: 1770.75, maximum: 1772.01
2022-03-23 15:50:06,217 [INFO] sarracenia.flowcb.log on_housekeeping housekeeping
2022-03-23 15:50:08,470 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CPOX/2022-03-23-1919-CPOX-AUTO-minute-swob.xml
2022-03-23 15:50:08,470 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CVFN/2022-03-23-1919-CVFN-AUTO-minute-swob.xml
2022-03-23 15:50:08,471 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWZT/2022-03-23-1919-CWZT-AUTO-minute-swob.xml
2022-03-23 15:50:08,471 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CPUN/2022-03-23-1919-CPUN-AUTO-minute-swob.xml
2022-03-23 15:50:08,471 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1806.14 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CVYD/2022-03-23-1919-CVYD-AUTO-minute-swob.xml
2022-03-23 15:50:08,471 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1806.14 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CTZN/2022-03-23-1919-CTZN-AUTO-minute-swob.xml
2022-03-23 15:50:08,471 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CZGH/2022-03-23-1919-CZGH-AUTO-minute-swob.xml
2022-03-23 15:50:08,472 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWIW/2022-03-23-1919-CWIW-AUTO-minute-swob.xml
2022-03-23 15:50:08,472 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1806.14 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CZKA/2022-03-23-1919-CZKA-AUTO-minute-swob.xml
2022-03-23 15:50:08,472 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1806.14 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CXET/2022-03-23-1919-CXET-AUTO-minute-swob.xml
2022-03-23 15:50:08,472 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CTPM/2022-03-23-1919-CTPM-AUTO-minute-swob.xml
2022-03-23 15:50:08,472 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.66 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CPQD/2022-03-23-1919-CPQD-AUTO-minute-swob.xml
2022-03-23 15:50:08,472 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1806.14 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/COGE/2022-03-23-1919-COGE-AUTO-minute-swob.xml
2022-03-23 15:50:08,473 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.65 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CTUX/2022-03-23-1919-CTUX-AUTO-minute-swob.xml
2022-03-23 15:50:08,473 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1806.14 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CASF/2022-03-23-1919-CASF-AUTO-minute-swob.xml
2022-03-23 15:50:08,473 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1805.65 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWNH/2022-03-23-1919-CWNH-AUTO-minute-swob.xml
2022-03-23 15:50:08,473 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1806.14 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CPYQ/2022-03-23-1919-CPYQ-AUTO-minute-swob.xml

```

then it was intermittently recovering for a while until this weirdness happenned:

```

2022-03-23 17:11:13,001 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1576.31 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWCO/2022-03-23-2044-CWCO-AUTO-minute-swob.xml
2022-03-23 17:11:13,001 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1576.35 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWPZ/2022-03-23-2044-CWPZ-AUTO-minute-swob.xml
2022-03-23 17:11:13,001 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1576.31 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CXFV/2022-03-23-2044-CXFV-AUTO-minute-swob.xml
2022-03-23 17:11:13,001 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1576.35 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWVT/2022-03-23-2044-CWVT-AUTO-minute-swob.xml
2022-03-23 17:22:02,796 [ERROR] sarracenia.transfer read_writelocal util/writelocal mismatched file length writing 2022-03-23-2039-CWXA-AUTO-minute-swob.xml. Message said to expect 8564 bytes.  Got 8563 bytes.
2022-03-23 17:22:02,796 [ERROR] sarracenia.flow download incomplete download only 8563 of expected 8564 bytes for 2022-03-23-2039-CWXA-AUTO-minute-swob.xml
2022-03-23 17:22:02,796 [INFO] sarracenia.flow do_download attempt 1 failed to download https://dd5.weather.gc.ca//observations/swob-ml/20220323/CWXA/2022-03-23-2039-CWXA-AUTO-minute-swob.xml to /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CWXA/2022-03-23-2039-CWXA-AUTO-minute-swob.xml
2022-03-23 17:22:02,796 [WARNING] sarracenia.flow do_download downloading again, attempt 2
2022-03-23 17:22:10,904 [ERROR] sarracenia.transfer read_writelocal util/writelocal mismatched file length writing 2022-03-23-2039-CWXA-AUTO-minute-swob.xml. Message said to expect 8564 bytes.  Got 8563 bytes.
2022-03-23 17:22:10,904 [ERROR] sarracenia.flow download incomplete download only 8563 of expected 8564 bytes for 2022-03-23-2039-CWXA-AUTO-minute-swob.xml
2022-03-23 17:22:10,904 [INFO] sarracenia.flow do_download attempt 2 failed to download https://dd5.weather.gc.ca//observations/swob-ml/20220323/CWXA/2022-03-23-2039-CWXA-AUTO-minute-swob.xml to /home/rolf/weather/swobMirror/observations/swob-ml/20220323/CWXA/2022-03-23-2039-CWXA-AUTO-minute-swob.xml
2022-03-23 17:22:10,904 [WARNING] sarracenia.flow do_download downloading again, attempt 3
2022-03-23 17:22:19,008 [ERROR] sarracenia.transfer read_writelocal util/writelocal mismatched file length writing 2022-03-23-2039-CWXA-AUTO-minute-swob.xml. Message said to expect 8564 bytes.  Got 8563 bytes.
2022-03-23 17:22:19,008 [ERROR] sarracenia.flow download incomplete download only 8563 of expected 8564 bytes for 2022-03-23-2039-CWXA-AUTO-minute-swob.xml
2022-03-23 17:22:19,008 [INFO] sarracenia.flow do_download attempt 3 failed to download https://dd5.weather.gc.ca//observations/swob-ml/20220323/CWXA/2022-03-23-2039-CWXA-AUTO-minute-swob.xml to /home/rolf/weather/swobMirror

```

after recovering from that, then it was suddenly only 5 minutes behind, instead of 30:

```

022-03-23 17:25:27,174 [INFO] sarracenia.diskqueue on_housekeeping post_retry_001 on_housekeeping
2022-03-23 17:25:27,174 [INFO] sarracenia.diskqueue on_housekeeping No retry in list
2022-03-23 17:25:27,174 [INFO] sarracenia.diskqueue on_housekeeping on_housekeeping elapse 0.000157
2022-03-23 17:25:27,174 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current Memory cpu_times: user=3025.45 system=575.89 elapse=17455928.9
2022-03-23 17:25:27,174 [INFO] sarracenia.flowcb.housekeeping.resources on_housekeeping Current Memory usage: 75.3 MiB / 224.3 MiB = 33.56%
2022-03-23 17:25:27,174 [INFO] sarracenia.flowcb.log housekeeping_stats messages_received: 100, accepted: 97, rejected: 3 ( bytes: 2134 )
2022-03-23 17:25:27,174 [INFO] sarracenia.flowcb.log housekeeping_stats files transferred: 93, cumulative bytes of data: 858293
2022-03-23 17:25:27,174 [INFO] sarracenia.flowcb.log housekeeping_stats lag: average: 1586.41, maximum: 1613.64
2022-03-23 17:25:27,174 [INFO] sarracenia.flowcb.log on_housekeeping housekeeping
2022-03-23 17:25:29,284 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 297.36 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWXA/2022-03-23-2020-CWXA-AUTO-minute-swob.xml
2022-03-23 17:25:29,284 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 297.36 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CACQ/2022-03-23-2043-CACQ-AUTO-minute-swob.xml
2022-03-23 17:25:29,284 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 296.33 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CVYD/2022-03-23-2043-CVYD-AUTO-minute-swob.xml
2022-03-23 17:25:29,284 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 296.33 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CABF/2022-03-23-2043-CABF-AUTO-minute-swob.xml
2022-03-23 17:25:29,284 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 296.33 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWPN/2022-03-23-2043-CWPN-AUTO-minute-swob.xml
2022-03-23 17:25:29,284 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 296.33 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWJV/2022-03-23-2044-CWJV-AUTO-minute-swob.xml

```

but the lag lengthened again... gradually... and the transfers went slowly, so it fell behind.

```

2022-03-23 17:51:45,770 [INFO] sarracenia.flowcb.log housekeeping_stats lag: average: 956.50, maximum: 962.80
2022-03-23 17:51:45,770 [INFO] sarracenia.flowcb.log on_housekeeping housekeeping
2022-03-23 17:51:48,669 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1593.56 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CXKI/2022-03-23-2121-CXKI-AUTO-minute-swob.xml
2022-03-23 17:51:48,669 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1592.77 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWFF/2022-03-23-2121-CWFF-AUTO-minute-swob.xml
2022-03-23 17:51:48,669 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1593.56 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CXRB/2022-03-23-2121-CXRB-AUTO-minute-swob.xml

```

lag wandered around again, and built up to 30 minutes (1800 seconds) again.  peaked at 1900.
and gradually recovered to <10... then: weird grouping of old messages.   My guess is that when it was restarted.
when the re-connection happenned, it resent some old messages that had never been acknowledged:


```
022-03-23 20:26:09,926 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220324/CNCD/2022-03-24-0025-CNCD-AUTO-minute-swob.xml
2022-03-23 20:26:09,926 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220324/CWCH/2022-03-24-0025-CWCH-AUTO-minute-swob.xml
2022-03-23 20:26:09,926 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220324/CTPM/2022-03-24-0025-CTPM-AUTO-minute-swob.xml
2022-03-23 20:26:09,948 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 13274.32 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWXA/2022-03-23-2039-CWXA-AUTO-minute-swob.xml
2022-03-23 20:26:09,948 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 13274.32 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWXA/2022-03-23-2035-CWXA-AUTO-minute-swob.xml
2022-03-23 20:26:09,948 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 13274.31 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWXA/2022-03-23-2033-CWXA-AUTO-minute-swob.xml
2022-03-23 20:26:09,948 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 13273.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWXA/2022-03-23-2043-CWXA-AUTO-minute-swob.xml
2022-03-23 20:26:10,046 [ERROR] sarracenia.transfer read_writelocal util/writelocal mismatched file length writing 2022-03-23-2039-CWXA-AUTO-minute-swob.xml. Message said to expect 8564 bytes.  Got 8563 bytes.
2022-03-23 20:26:10,046 [ERROR] sarracenia.flow download incomplete download only 8563 of expected 8564 bytes for 2022-03-23-2039-CWXA-AUTO-minute-swob.xml

```
followed immediately by <2 second lag...

```
2022-03-23 20:26:13,498 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.28 ) https://dd5.weather.gc.ca /observations/swob-ml/20220324/CWLJ/2022-03-24-0025-CWLJ-AUTO-minute-swob.xml
2022-03-23 20:26:13,498 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.27 ) https://dd5.weather.gc.ca /observations/swob-ml/20220324/CZEV/2022-03-24-0025-CZEV-AUTO-minute-swob.xml
2022-03-23 20:26:13,498 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.27 ) https://dd5.weather.gc.ca /observations/swob-ml/20220324/CNEK/2022-03-24-0025-CNEK-AUTO-minute-swob.xml
2022-03-23 20:26:13,498 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.27 ) https://dd5.weather.gc.ca /observations/swob-ml/20220324/CWJR/2022-03-24-0025-CWJR-AUTO-minute-swob.xml
2022-03-23 20:26:13,499 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220324/CZGH/2022-03-24-0025-CZGH-AUTO-minute-swob.xml
2022-03-23 20:26:13,499 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220324/CVFN/2022-03-24-0025-CVFN-AUTO-minute-swob.xml
2022-03-23 20:26:13,499 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220324/CMSI/2022-03-24-0025-CMSI-AUTO-minute-swob.xml
```

**Comment by endlisnis** [comment]
> It doesn't look like it logged anything on shutdown or startup

When I restarted it, none of the instances responded to the TERM requests.  They ended up getting killed after a while.

This is the timing of the restart:
`2022-03-23 20:19:25,571 [INFO] root stop_signal signal 15 received`

**Comment by endlisnis** [comment]
I also noticed that sometimes it saves partial downloaded files.  This is really brutal for my application.  I have a periodic task that runs once-per-hour that tries to poll for files that were missed; but since PART of these files were downloaded, I can't easily tell that I should try to download it again.

I'm having to check the file extension, and if it's .xml, then I have to see if ">" is the last character in the file.  If it is not, then assume this is corrupt and download it again.

For my application, it would be much better if partial downloads were thrown away.

**Comment by matthewdarwin** [comment]
I wrote my own downloader rather than the built-in one.

I have a custom pluging where serra just sends notification to my downloader.  My downloader

- re-orders the queue of files based on importance (some times get too many notifications and DD team doesn't like me downloading that many files at once, so need to slow it down)
- downloads the file to a temporary location
- if the file is of type xml, validates that the xml document is valid
- then moves the file into place atomically on a journaling filesystem to avoid corruption
- the additional background process that checks for missing files (run directory listing on all directories on DD and compare to files I have) just has to check what is missing as there are no corrupt files in the directory at this point

**Comment by petersilva** [comment]
fwiw @matthewdarwin ... you can do all that built-in downloader:
```
inflight tmp/

```
would download the file to a subdirectory named tmp instead ot files named .tmp and then move them when complete.
```
acceptSizeWrong off

```
will have it fail the downloads of files when the size is wrong, and it will retry them automatically.
```
instances 4

```
you can control how many downloaders operate in parellel with instances.  we have cases with 40 or more processes munching on the same queue. The acceptSizeWrong is perhaps only in sr3, but the others have been there since 2018 at least.

To prioritize things, configure a separate subscriptions for the high priority stuff vs. the others (two separate configurations) so that there is a high priority queue and and another queue for everything else.

**Comment by petersilva** [comment]
@endlisnis the option:

for the wrong size giving you headaches... short version add a line like this to your config:
```

acceptSizeWrong off

```
https://metpx.github.io/sarracenia/Reference/sr3_options.7.html#acceptsizewrong-boolean-default-false

**Comment by matthewdarwin** [comment]
Thanks for the hints. I need to prioritize downloads amongst the background scraper jobs too, which don't run via sarra.  If internet connection is down for a few hours, there are many files to download and catch up.

**Comment by petersilva** [comment]
hey @matthewdarwin if the download CLI isn't giving any value to you and you are using custom code anyways, you might want to use the (new) python API, and not bother with the CLI at all.  Demonstration here:

 * https://metpx.github.io/sarracenia/Tutorials/4_api_moth_sub_demo.html

**Comment by endlisnis** [comment]
`acceptSizeWrong off` should be the default in v3.  For 99% of the use cases I can think of, keeping partial files is wrong and complicated.

**Comment by petersilva** [comment]
@endlisnis yes... looking at the code... I'm confused... because it already is the default... why it would not retry in that case is puzzling... looking.... wow... it did reject the file... it wasn't happy. it just failed to queue it up for retry... but it got it half-right, in that the reception was never acknowledged, so that later: 
```

2022-03-23 20:26:09,948 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 13274.31 ) https://dd5.weather.gc.ca /observations/swob-ml/20220323/CWXA/2022-03-23-2033-CWXA-AUTO-minute-swob.xml

```
so it's a bug where it has kind of half-detected that the transfer failed... AMQP decided to resend the
message when re-connection happenned. why didn't it get queued for retry?
oh... it did:
```
2022-03-23 19:38:39,792 [INFO] sarracenia.diskqueue on_housekeeping Number of messages in retry list 4
2022-03-23 19:38:39,793 [INFO] sarracenia.diskqueue on_housekeeping on_housekeeping elapse 0.006980
2022-03-23 19:38:39,793 [INFO] sarracenia.diskqueue on_housekeeping post_retry_001 on_housekeeping
2022-03-23 19:38:39,793 [INFO] sarracenia.diskqueue on_housekeeping No retry in list

``
so getting on the retry list is ok.  but then it prints 'No Retry in list'... so ... it's like it cannot read from it.
Hey @reidsunderland this looks like the same case you were seeing in the flow test?
it's real!  stuff is getting put on the retry list, but fetching from it isn't working in at least one real case. 

* "get" from the retry queue seems to be failing.
* even if the transfer failed, it should have acknowledged the message, as it should be on the retry queue,
  and download will be re-attempted later.

So there is a problem in the retry logic... we will work on that.  I think @reidsunderland might already on it in the context of another issue.

**Comment by reidsunderland** [comment]
Yes! It seems to be related to, or the same problem that I'm working on for #466.

**Comment by endlisnis** [comment]
I continue to have daily problems with my feeds falling behind (often lag exceeding 400 seconds, sometimes getting up near 1500 seconds).  Do you need any more logs from me?

**Comment by petersilva** [comment]
#488 might fix this.

**Comment by endlisnis** [comment]
I have the fix for #488, but I (again) experienced a slowdown today.
[subscribe_rolf_01.log.gz](https://github.com/MetPX/sarracenia/files/8365742/subscribe_rolf_01.log.gz)

**Comment by petersilva** [comment]
Analyzing this one.
-------------------------

was going up and down, as bad as 5 minutes, but usually only a few seconds, until:

```

2022-03-28 02:19:19,697 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CXWG/2022-03-28-0618-CXWG-AUTO-minute-swob.xml
2022-03-28 02:19:19,697 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWJR/2022-03-28-0618-CWJR-AUTO-minute-swob.xml
2022-03-28 02:19:19,697 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CTPQ/2022-03-28-0618-CTPQ-AUTO-minute-swob.xml
2022-03-28 02:19:19,697 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1.26 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWLJ/2022-03-28-0617-CWLJ-AUTO-minute-swob.xml
2022-03-28 03:25:28,486 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220328/CZPK/2022-03-28-0601-CZPK-AUTO-minute-swob.xml
2022-03-28 03:25:28,487 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220328/CARP/2022-03-28-0618-CARP-AUTO-minute-swob.xml
20
```
then it hung for an hour or so...

``` 
2022-03-28 03:25:28,495 [INFO] sarracenia.flowcb.log stats lag: average: 4.07, maximum: 19.37
2022-03-28 03:25:28,495 [INFO] sarracenia.flowcb.log on_housekeeping housekeeping
2022-03-28 03:25:31,292 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1603.51 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CVGE/2022-03-28-0653-CVGE-AUTO-minute-swob.xml
2022-03-28 03:25:31,293 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1603.51 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWXB/2022-03-28-0658-CWXB-AUTO-minute-swob.xml
2022-03-28 03:25:31,293 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1597.86 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWPU/2022-03-28-0658-CWPU-AUTO-minute-swob.xml
2022-03-28 03:25:31,293 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1597.86 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/COGE/2022-03-28-0658-COGE-AUTO-minute-swob.xml
2022-03-28 03:25:31,293 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1598.39 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CVRA/2022-03-28-0658-CVRA-AUTO-minute-swob.xml

```

caught up in 4 minutes after restoral:

```
2022-03-28 03:39:55,068 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 0.50 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CVBB/2022-03-28-0739-CVBB-AUTO-minute-swob.xml
2022-03-28 03:39:55,069 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 0.49 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWVQ/2022-03-28-0739-CWVQ-AUTO-minute-swob.xml
2022-03-28 03:39:55,069 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 0.49 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWUS/2022-03-28-0739-CWUS-AUTO-minute-swob.xml
2022-03-28 03:39:55,590 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220328/CVBB/2022-03-28-0739-CVBB-AUTO-minute-swob.xml
```

The symptom is different.  It doesn't show any error message, but just hangs.  did you restart around 3:25 ?
perhaps a cron job?
anyways, then it was good for about 6 hours, and then it gradually slowed down:

```
2022-03-28 09:40:50,470 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 8.07 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWCT/2022-03-28-1339-CWCT-AUTO-minute-swob.xml
2022-03-28 09:40:50,471 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 7.91 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWTE/2022-03-28-1339-CWTE-AUTO-minute-swob.xml
2022-03-28 09:40:50,471 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 8.05 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CNZS/2022-03-28-1339-CNZS-AUTO-minute-swob.xml
2022-03-28 09:47:37,401 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220328/CMGZ/2022-03-28-1339-CMGZ-AUTO-minute-swob.xml
.
.
.
2022-03-28 09:47:39,509 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 400.76 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWSZ/2022-03-28-1340-CWSZ-AUTO-minute-swob.xml
2022-03-28 09:47:39,509 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 400.75 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CXCA/2022-03-28-1340-CXCA-AUTO-minute-swob.xml

... then a 13 minute delay while downloading occurs... this 13 minutes represents the time before start of download and completion for the batch of files. accepted above.

2022-03-28 09:47:39,512 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 397.56 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWEI/2022-03-28-1340-CWEI-AUTO-minute-swob.xml
2022-03-28 10:01:12,261 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220328/CWSZ/2022-03-28-1340-CWSZ-AUTO-minute-swob.xml

.
.
.

... in the same second, it receives the next batch of messages:
2022-03-28 10:01:12,263 [INFO] sarracenia.flowcb.log after_work downloaded ok: /home/rolf/weather/swobMirror/observations/swob-ml/20220328/CWEI/2022-03-28-1340-CWEI-AUTO-minute-swob.xml
2022-03-28 10:01:12,263 [INFO] sarracenia.flow run on_housekeeping pid: 3385964 subscribe/rolf instance: 1
2022-03-28 10:01:12,263 [INFO] sarracenia.flowcb.gather.message on_housekeeping messages: good: 100 bad: 0 bytes: 11.9 KiB average: 121 Bytes
... now it delayed by about 18 minutes...
2022-03-28 10:01:12,264 [INFO] sarracenia.flowcb.log on_housekeeping housekeeping
2022-03-28 10:01:14,423 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1136.29 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CWLV/2022-03-28-1341-CWLV-AUTO-minute-swob.xml
2022-03-28 10:01:14,423 [INFO] sarracenia.flowcb.log after_accept accepted: (lag: 1136.29 ) https://dd5.weather.gc.ca /observations/swob-ml/20220328/CERM/2022-03-28-1341-CERM-AUTO-minute-swob.xml

```
so... that doesn't look like any client-side error, but rather internet congestion...  It could be on the our end or the your end, or somewhere in between. hmm... looking at  stats... yeah there is heavy load, but it does not look saturated... 
Are you pulling down a lot of other stuff at the same time (maybe weather model outputs?) that could compete with this?

**Comment by petersilva** [comment]
oh you can also set *batch 1* if you want to see the delay accumulate more directly.

**Comment by endlisnis** [comment]
That computer is pretty idle during the day.  I only do significant downloading once per hour at *:15 when I manually sync up with the SWOB directories.  If the AMQP client is far behind, that might be a bunch of downloading.

I do share my home internet with my work connection.  I'm often doing Teams video meetings on it, but it's 100M service, so it shouldn't be THAT bad.  Also, I would expect the clients to catch up after those sessions.

I restarted it at 15:34:43.  That's the event that I was trying to show.  I guess the log file I posted didn't show it catch up quickly after that restart.  It was back to ~5s lag within 3 minutes.

The reason I don't think it's just congestion is that it caught up really fast after I restarted the service.

**Comment by petersilva** [comment]
swob is tiny... it's should not be that big a deal.  no... I was thinking piles of model outputs, or satellite imagery or something...
teams is supposed to be < 2mbs... and you say a restart fixed it... hmm... odd.

**Comment by petersilva** [comment]
so there is a byteRateMax feature to do bandwidth limiting... if that is being triggerred inappropriately, then with the latest commit, you should see an info message
```

  [INFO] ... throttle exceeded byteRateMax X sleeping for y 

```
I doubt that's it, but might as well.

**Comment by endlisnis** [comment]
I just updated and restarted.  Let's see what tomorrow looks like...

**Comment by endlisnis** [comment]
Another day, another slowdown.  After it was more than 1000 seconds behind and not catching up, I restarted it at "21:10:06" and it quickly caught up.
[subscribe_rolf_01.log.gz](https://github.com/MetPX/sarracenia/files/8393996/subscribe_rolf_01.log.gz)

**Comment by petersilva** [comment]
How many instances are you running?  I think 5 would be a good number.  if you run only 1, then all the downloads are sequential, if you have multiple instances then they overlap (from the various instances.) it might just accumulating delay from making everything sequential.   the log above doesn't look that bad, it gets slow and catches up multiple times during the day... it isn't clear that it would not have caught up again.  hmm... but it would be good to see what happens if there is >1 instances.
If it's already >1, then I'll need to think of something else

```
instances 5
```

**Comment by endlisnis** [comment]
I'm running 4 instances.

If you look at just before I reset it.  It had sat for literally 12 minutes with no output.  The same was true for the other 3 instances.  Nobody was downloading anything.

**Comment by petersilva** [comment]
All of them stopped at the same time? It's not like instances communicate with one another... they are just independent processes.  It's hard to see how a software problem would affect four processes identically after hours of downloading different stuff.  It's starting to look like the broker is involved here, rather than the client software... hmm...

**Comment by petersilva** [comment]
I suggest 

```
   batch 5
```
You see these long gaps because the the actual downloader is not printing anything, it receives a batch of messages,  accepts them, then it downloades them all, then after it has finished downloading the entire batch, it then prints download messages.  so that ten minute gap is while it is downloading all the stuff accepted above.  if you reduce the batch size you will see quicker alternation between accept and download. and it might not appear stalled.

**Comment by endlisnis** [comment]
I changed my config to "batch 1" just to be pedantic.

I took a more detailed look at the logs for March 31st.  They did not all stop at the same time.
(1) stopped at 20:58:29.
(2) stopped at 21:02:58.
(3) stopped at 21:07:34.
(4) stopped at 20:59:50.

That's scattered enough for me to believe the downloader is just slowing down (as you said).

I still think that something is going on here.  Sometimes, the downloading almost stops (I guess it's technically still going, but if it takes more than 10 minutes to download a batch, I feel that's effectively stopped).  But if I restart it, it catches up very quickly (indicating that it's not my network connection that's slowing it down).

It will be interesting to see what the slowdowns look like with "batch 1".

**Comment by petersilva** [comment]
It's clustered within the same 9 minute period... if they each instance had a certain amount of message pre-fetched, then this kind of spread looks normal, to me.  So far, I'm thinking the broker is slowing down... but will be interesting to see.

**Comment by endlisnis** [comment]
OK, I had another slowdown today after I switched to "batch 1" and I noticed an interesting pattern.
[subscribe_rolf_01.log.gz](https://github.com/MetPX/sarracenia/files/8403592/subscribe_rolf_01.log.gz)

I assume that the "after_accept accepted" happens before the download and "after_work downloaded" happens after the download, and I *also* assume that the time between those lines is approximately the download time.

Given those assumptions, around "19:25:35", the time to download each file changes from ~100ms to ~8000ms.  And it stays at very close to 8000ms for more than an hour, until I perform a reset at "20:36:31".  After that, it immediately goes back to ~100ms and very quickly catches up.

**Comment by endlisnis** [comment]
I had another slowdown today.  Same signature.  The download time suddenly jumped up to ~8s per file.  It remained there for about 2 hours until I noticed it.  After I reset the service, it quickly caught back up.

**Comment by petersilva** [comment]
nothing new to say... hmm... It would be good to note the time of day when you restart... if there is no difference in the logs based on what you have already provided... I'll look into just reproducing the whole flow here...
ah... idea occurs:

question... are you still writing file name from a named pipe using a plugin?  I remember a while ago... I think you had forgotten about it and the plugins was still there, but you weren't using it anymore?  It would be good to remove that if you don't need it, and see if this slow down goes away...

**Comment by endlisnis** [comment]
Yesterday, on,
instance 1: the downloads jumped to ~8s at "19:11:07" and I initiated a reset at "21:09:29".
instance 2: the downloads jumped to ~8s at "19:11:06".
instance 3: "19:11:07"
instance 4: "19:11:06"

So, clearly, all 4 instances are experiencing some problem at almost exactly the same second -- and just never clearing it.

I am still using the named pipe.  I write the filename to the pipe, but I actually only use the IATA code from the filename on the other end.  I can't remove that plugin without using some OTHER way of notifying my program of a new file ready for processing.

**Comment by petersilva** [comment]
Aha! a pattern... that is encouraging! That's midnight z in the winter, but only 23h in the summer.

Um do you have a cron job running that might be archiving on your disk or something competing with the downloader for io? 

You could put in a cron job at  19:15 to restart it automagically as a work-around... but it would be nice to actually figure out.

it's probably not the named pipe thing, or if it is... maybe that thing dies at 19:00 and then the pipe fills up and weird slow downs happen, then when you restart, it restarts the other thing, and empties the pipe?

but I don't know what sort of other processing you are doing, but we could do a plugin to invoke it directly from sarra, rather than doing IPC?  but I don't think that's the problem based on the timing info.

**Comment by endlisnis** [comment]
It happened at 19:25 one day, and 19:11 3 days later.  That is interesting.

I do have an archiving task, but it runs at 02:00 each day.  The pipe will stall for about 40 minutes during that archiving, but (oddly enough), that never seems to bother sr3; it recovers from that just fine.

I do run hourly tasks at 15 minutes after each hour, but that should not really bother sr3 (in theory).

When I restart sr3, it does NOT restart the other end of the pipe, so I don't see how restarting sr3 would somehow recover both ends of that named pipe.  The other end of the pipe is idle during those slowdown periods.

There are certainly times when my scripts fail to read from the pipe (they sometimes block for hours and need some human interaction to fix them up again).  But when that happens, sr3 just blocks forever (which is fine -- it is NOT what we are seeing here).  I would actually love a way to have a *much* deeper named pipe (think hundreds of megs), but I haven't put too much thought into it.

**Comment by petersilva** [comment]
for pipe... we could  just write to an hourly file, and your reader could just stay an hour behind all the time..
so we write a new file every hour, and your reader reads all the files except the current one...

but not solving the issue...

**Comment by petersilva** [comment]
you don't happen to know if it happened at an hour later during the winter, do you? that would indicate a UTC cron job on our side perhaps.

**Comment by endlisnis** [comment]
I run into trouble when I use hpfx: I use `mirror True`, and that ends up pushing files to a different path on my local setup.  Is there a way to strip off some path parts for the mirror path?

**Comment by petersilva** [comment]
```
strip 2 
```

strip first two elements of the relPath? 

https://metpx.github.io/sarracenia/Reference/sr3_options.7.html?highlight=strip#strip-count-regexp-default-0

**Comment by petersilva** [comment]
```

pip install -e .

```

again... sometimes things get messed up?

**Comment by petersilva** [comment]
can you cat your config file? this stuff works fine for the configs I have...

**Comment by petersilva** [comment]
perhaps try adding to ~/.config/sr3/credentials.conf:
```

amqps://anonymous:anonymous@hpfx.collab.science.gc.ca

```

**Comment by endlisnis** [comment]
```
rolf@endlisnis11 ~/weather
$ cat ~/.config/sr3/subscribe/rolf.conf
broker amqps://anonymous:anonymous@dd.weather.gc.ca/

queue_name q_anonymous.rolf.20211222
instances 4

directory /home/rolf/weather/swobMirror/
# All stations
topicPrefix v02.post
subtopic observations.swob-ml.#
batch 1

mirror True
reject .*/partners/.*
reject .*/moored-buoys/.*
accept .*
expire 4h
acceptSizeWrong off
inflight tmp/

rxpipe_name /home/rolf/weather/rxpipe
flowCallback rxpipe_gzip.RxPipe_gzip

rolf@endlisnis11 ~/weather
$ ~/.local/bin/sr3 stop subscribe/rolf
Traceback (most recent call last):
  File "/home/rolf/.local/bin/sr3", line 33, in <module>
    sys.exit(load_entry_point('metpx-sr3', 'console_scripts', 'sr3')())
  File "/home/rolf/weather/sr3/sarracenia/sr.py", line 2019, in main
    gs = sr_GlobalState(cfg, cfg.configurations)
  File "/home/rolf/weather/sr3/sarracenia/sr.py", line 1020, in __init__
    self._read_configs()
  File "/home/rolf/weather/sr3/sarracenia/sr.py", line 307, in _read_configs
    cfgbody.fill_missing_options(c, cfg)
  File "/home/rolf/weather/sr3/sarracenia/config.py", line 1498, in fill_missing_options
    queuefile += os.sep + component + '.' + cfg + '.' + self.broker.url.username
TypeError: can only concatenate str (not "NoneType") to str
```

**Comment by endlisnis** [comment]
I had to change the broker line to:
`broker  https://anonymous:anonymous@hpfx.collab.science.gc.ca`
To make it work.

**Comment by endlisnis** [comment]
And then I got:
```
2022-04-07 19:07:58,798 [CRITICAL] sarracenia.moth ProtocolPresent Protocol scheme https unsupported for communications with message brokers
```

**Comment by endlisnis** [comment]
I got it working by switching to `broker  amqps://anonymous:anonymous@hpfx.collab.science.gc.ca`

**Comment by endlisnis** [comment]
A problem that got me very confused about this exception was that even though I was only trying to stop the "rolf" config, there was a problem with the "hpfx" config file which was causing the python to crash.  I was not expecting the script to fail on some OTHER config file when trying to stop one specific service.  I had to add some debug code into `config.py` to figure out what was going on.

**Comment by endlisnis** [comment]
[subscribe_hpfx_01.log.gz](https://github.com/MetPX/sarracenia/files/8447551/subscribe_hpfx_01.log.gz)


After switching over to hpfx, it ran for a little more than 30 minutes before getting stuck (in a different, but familiar way this time).

At "19:45:03", it started getting "[Errno -2] Name or service not known", and never recovered.  All the instances started at the exact same time.  I wouldn't be surprised to find out that there was a momentary failure in DNS, but it recovered very quickly.

At "20:01:52", I noticed the problem and restarted the service.  It very quickly caught up.

**Comment by petersilva** [comment]
* the confusion part... yeah.... sr3 reads the entire ~/.config/sr3 directory so you can do group actionsl like start a poll four 
subscribers and shovel in one command.  It is preparatory for the future where it can make consistency checks of large configurations (like every exchange has a consumer, just by reading the bindings of other configs) build a map of a pump.
in our big configurations, we have several hundred configs active. It's puzzling, not sure what to do about it.

* the broker error message is normal.. yup we can't talk to http brokers, only amqp and mqtt .  

* I reproduced your problem will try to deal with it better to get a better error message.

**Comment by endlisnis** [comment]
>  It's puzzling, not sure what to do about it.

Why not only read in all of the configs when you need to make the consistency checks?  When I'm asking to stop a specific service, I assume there's no point in doing global consistency checks.

Or, if there is a need to do global consistency checks even when stopping an individual service, you could warn about configs that are not parsable, but still continue on.

**Comment by endlisnis** [comment]
Well, I think I've figured out what triggers the 8-second download times.
I'm using VPN software (Cisco AnyConnect).  When the VPN disconnects, the sr3 instances start getting this 8 second delay on each file.
If I re-connect to the VPN (without restarting sr3), then the downloading goes back to normal and catches up.

I've tried this a few times now, and it's reproducible.

**Comment by petersilva** [comment]
Wow! That's great!  What's puzzling is why restarting sr3 ever worked?  when the vpn goes off, does anything else get slow? dns lookups? ping to google or something?

**Comment by endlisnis** [comment]
Oh, to clarify:  Disconnecting from the VPN slows down sr3.  Reconnecting to the VPN restores sr3 speed.  Restarting sr3 after the VPN disconnect also restores sr3 speed.

No other programs slow down on my system after the disconnect.  Some part of sr3 is maintaining state that causes the delay when the VPN disconnects.

I have a feeling that this is going to turn out to be a failed DNS lookup, or trying to route the request across a non-existing network interface.  The VPN tool does inject 2 extra DNS servers into `/etc/resolv.conf`.

Now, the funny thing is that I had no trouble reproducing this last night.  3 times in a row it would start having trouble exactly when my VPN disconnected, but now (this morning) I can't reproduce it.

I wonder if the VPN has to be connected for a while for sr3 to cache information about the alternate network.

**Comment by petersilva** [comment]
sr3 is a naive user of dns.  Reviewing source code,
* to connect to an AMQP broker, sr3 hands the hostname to the amqp library which does the lookup.
* to connect to a webserver, urllib.request.urlopen...just straight python, no smarts at all.

thing is, most browsers have built-in dns caching, so you cannot test that way.  you need to ping or lookup a host from the command line.  On Linux, there are things like nscd, and systemd-resolved.service that have rendered dns caching incomprehensible. These days I just test... and then debug from how it actually works, rather than trying to guess.

when logging in via SSL there might be some additional checks... as an A:B test, you could try adding:
```

tlsRigour lax

```
to see if the timing changes at all... it might be tls doing reverse lookup or something?

**Comment by petersilva** [comment]
if runing on linux... when sr3 is slow try:
```
    host -C www.cbc.ca

```
to get a more elaborate host lookup (involving consistency checks.)

**Comment by endlisnis** [comment]
That command outputs literally nothing when I run it on my linux machine.

```
rolf@endlisnis11 ~/weather/sr3
$ host -C www.cbc.ca

rolf@endlisnis11 ~/weather/sr3
$ host -C www.google.com

```

**Comment by endlisnis** [comment]
> sr3 is a naive user of dns. Reviewing source code,
> thing is, most browsers have built-in dns caching, so you cannot test that way. you need to ping or lookup a host from the command line. On Linux, there are things like nscd, and systemd-resolved.service that have rendered dns caching incomprehensible. These days I just test... and then debug from how it actually works, rather than trying to guess.

I have tried using "curl" to download the exact url of one of the swob xml files while sr3 was in slow mode.  It completed in 114ms.  I assume if there is any DNS caching going on that it's probably behaving the same in sr3 as it is in curl.

Previously, you mentioned something about re-using connections (this was for the bug where it would fail once, and then keep failing forever, the bug you already fixed).  Could it be related to that type of connection re-use?  Maybe the routing table changes when the VPN disconnects which causes old connections to have lots of packet loss or something.  The funny thing is that it does EVENTUALLY work (after ~8s).

**Comment by petersilva** [comment]
* if the connection is established, there  is no dns lookup... the scope of connections is controlled by *batch* size.  after *batch* transfers with the same connection, it tears it down, and opens a new one.  It doesn't match with dns as culprit.  maybe there is a bug and sr3 is opening a new connection every time? hmm...
* so curl is fine. and it would certainly need to do a lookup at least once... hmm...

**Comment by petersilva** [comment]
the host -C thing... I had given the info based on the man page... but I have the same result as you... not useful.

**Comment by endlisnis** [comment]
When I use a batch size of larger than 1, the time for the whole batch is about 8× the batch size.  So, yeah, maybe this isn't DNS.

**Comment by endlisnis** [comment]
I was able to reproduce it tonight.  I tried to use the same circumstances as before (I'm not sure how many of these pieces actually matter):
(1) I leave my VPN connected for a few hours.
(2) I run sr3 during those hours.
(3) At around 19:15, I disable the VPN.

Sure enough, I was able to reproduce the problem.

While I was in that state, I ran "host www.cbc.ca" and even

```
$ host hpfx.collab.science.gc.ca
hpfx.collab.science.gc.ca has address 142.98.224.27
```

No changes when it's run after I restart the service.

**Comment by endlisnis** [comment]
Is there same debug that I can enable that would give more output about what it's doing during that 8-second time?

**Comment by endlisnis** [comment]
> for pipe... we could just write to an hourly file, and your reader could just stay an hour behind all the time.. so we write a new file every hour, and your reader reads all the files except the current one...

I actually came up with a better solution for this.  I'm including my updated plugin.  It uses an sql database instead of a named pipe.
[rxqueue_gzip.py.gz](https://github.com/MetPX/sarracenia/files/8459260/rxqueue_gzip.py.gz)

**Comment by petersilva** [comment]
Very Cool! I would accept a PR if you want it added to Sarracenia itself...
or you could just add a doc string giving distribution permissions compatible with GPLv2 (used by sarracenia) either way authorship needs to be indicated somehow.

**Comment by endlisnis** [comment]
The hpfx server configuration hit the same slowdown as the dd configuration was doing before when my VPN software disconnected this morning.

I restarted it at "11:27:55" and it quickly caught up.

Although I guess that hpfx allows deeper queues, because I'm normally limited to ~1700 seconds of data, whereas I got well over 5000s on hpfx.

[subscribe_hpfx_01.log.gz](https://github.com/MetPX/sarracenia/files/8466246/subscribe_hpfx_01.log.gz)

**Comment by petersilva** [comment]
OK so that eliminates server side as a problem.  The solution is now very likely within sr3 itself.  I will look at the download code more closely.

**Comment by endlisnis** [comment]
> Very Cool! I would accept a PR if you want it added to Sarracenia itself... or you could just add a doc string giving distribution permissions compatible with GPLv2 (used by sarracenia) either way authorship needs to be indicated somehow.

I just created a pull request: https://github.com/MetPX/sarracenia/pull/512

**Comment by endlisnis** [comment]
>  The solution is now very likely within sr3 itself. I will look at the download code more closely.

Any updates on this?  I still experience this slowdown almost every day when my VPN disconnects.

**Comment by petersilva** [comment]
no, have not had time to look at this...

**Comment by endlisnis** [comment]
Well, I was originally running the Linux version of Cisco AnyConnect, not the windows version.

But, I no longer run it anyway.  I migrated my SR3 installation to a raspberry pi, (for other reasons), so I can neither confirm nor deny that it's fixed.

**Comment by petersilva** [comment]
OK... well it does not look like we will make any further progress on this. closing it for now...
we can always re-open if it becomes relevant again.
